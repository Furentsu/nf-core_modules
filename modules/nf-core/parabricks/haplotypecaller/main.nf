process PARABRICKS_HAPLOTYPECALLER {
    tag "$meta.id"
    label 'process_high'

    container "nvcr.io/nvidia/clara/clara-parabricks:4.0.1-1"

    /*
    NOTE: Parabricks requires the files to be non-symlinked
    Do not change the stageInMode to soft linked! This is default on Nextflow.
    If you change this setting be careful.
    */
    stageInMode "copy"

    input:
    tuple val(meta), path(bam), path(bam_index), path(intervals)
    tuple val(meta2), path(fasta)

    output:
    tuple val(meta), path("*.vcf.gz"), emit: vcf
    path "versions.yml"              , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    // Exit if running this module with -profile conda / -profile mamba
    if (workflow.profile.tokenize(',').intersect(['conda', 'mamba']).size() >= 1) {
        error "Parabricks module does not support Conda. Please use Docker / Singularity / Podman instead."
    }
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    def interval_command = intervals ? intervals.collect{"--interval-file $it"}.join(' ') : ""
    def copy_index_command = bam_index ? "cp -L $bam_index `readlink -f $bam`.bai" : ""

    """
    # parabricks complains when index is not a regular file in the same directory as the bam
    # copy the index to this path.
    $copy_index_command

    pbrun \\
        haplotypecaller \\
        --ref $fasta \\
        --in-bam $bam \\
        $interval_command \\
        --out-variants ${prefix}.vcf.gz
        --num-threads $task.cpus \\
        --num-gpus $task.accellerator.request \\
        $args

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
            pbrun: \$(echo \$(pbrun version 2>&1) | sed 's/^Please.* //' )
    END_VERSIONS
    """

    stub:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    def interval_command = intervals ? intervals.collect{"--interval-file $it"}.join(' ') : ""

    """
    touch ${prefix}.vcf.gz

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
            pbrun: \$(echo \$(pbrun version 2>&1) | sed 's/^Please.* //' )
    END_VERSIONS
    """
}
