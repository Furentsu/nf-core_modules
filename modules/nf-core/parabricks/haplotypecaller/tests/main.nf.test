nextflow_process {

    name "Test Process PARABRICKS_HAPLOTYPECALLER"
    script "../main.nf"
    process "PARABRICKS_HAPLOTYPECALLER"

    tag "modules"
    tag "modules_nfcore"
    tag "parabricks"
    tag "parabricks/haplotypecaller"

    test("sarscov2 - bam - no_intervals") {
        config "./cp-nextflow.config"
        when {
            process {
                """
                input[0] = [
                            [ id:'test' ],
                            file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/bam/test.paired_end.sorted.bam', checkIfExists: true)
                ]
                input[1] = [
                            [ id:'index' ],
                            []
                ]
                input[2] = [
                            [ id:'intervals' ],
                            []
                ]
                input[3] = [
                            [ id:'reference' ],
                            file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true)
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    path(process.out.vcf[0][1]).vcf.variantsMD5,
                    process.out.versions
                ).match() }
            )
        }

    }

    test("sarscov2 - bam - intervals") {
        config "./cp-nextflow.config"
        when {
            process {
                """
                input[0] = [
                            [ id:'test' ],
                            file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/bam/test.paired_end.sorted.bam', checkIfExists: true)
                ]
                input[1] = [
                            [ id:'index' ],
                            file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/bam/test.paired_end.sorted.bam.bai', checkIfExists: true)
                ]
                input[2] = [
                            [ id:'intervals' ],
                            file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/bed/test.bed', checkIfExists: true)
                ]
                input[3] = [
                            [ id:'reference' ],
                            file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true)
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    path(process.out.vcf[0][1]).vcf.variantsMD5,
                    process.out.versions
                ).match() }
            )
        }

    }

    test("sarscov2 - bam - gvcf") {
        config "./nextflow.config"
        when {
            process {
                """
                input[0] = [
                            [ id:'test' ],
                            file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/bam/test.paired_end.sorted.bam', checkIfExists: true)
                ]
                input[1] = [
                            [ id:'index' ],
                            file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/bam/test.paired_end.sorted.bam.bai', checkIfExists: true)
                ]
                input[2] = [
                            [ id:'intervals' ],
                            file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/bed/test.bed', checkIfExists: true)
                ]
                input[3] = [
                            [ id:'reference' ],
                            file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true)
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    path(process.out.vcf[0][1]).vcf.variantsMD5,
                    process.out.versions
                ).match() }
            )
        }

    }

    test("sarscov2 - bam - stub") {
        config "./cp-nextflow.config"
        options "-stub"

        when {
            process {
                """
                input[0] = [
                            [ id:'test' ],
                            file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/bam/test.paired_end.sorted.bam', checkIfExists: true)
                ]
                input[1] = [
                            [ id:'index' ],
                            file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/bam/test.paired_end.sorted.bam.bai', checkIfExists: true)
                ]
                input[2] = [
                            [ id:'intervals' ],
                            file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/bed/test.bed', checkIfExists: true)
                ]
                input[3] = [
                            [ id:'reference' ],
                            file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true)
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

    test("sarscov2 - bam - gvcf - stub") {
        config "./nextflow.config"
        options "-stub"

        when {
            process {
                """
                input[0] = [
                            [ id:'test' ],
                            file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/bam/test.paired_end.sorted.bam', checkIfExists: true)
                ]
                input[1] = [
                            [ id:'index' ],
                            file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/bam/test.paired_end.sorted.bam.bai', checkIfExists: true)
                ]
                input[2] = [
                            [ id:'intervals' ],
                            file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/bed/test.bed', checkIfExists: true)
                ]
                input[3] = [
                            [ id:'reference' ],
                            file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true)
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

}
