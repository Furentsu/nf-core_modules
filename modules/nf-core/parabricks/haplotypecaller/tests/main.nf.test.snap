{
    "sarscov2 - bam - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test.vcf:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    "versions.yml:md5,cc6d48cc1e81c73a02d158459281b500"
                ],
                "vcf": [
                    [
                        {
                            "id": "test"
                        },
                        "test.vcf:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,cc6d48cc1e81c73a02d158459281b500"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.1",
            "nextflow": "24.10.0"
        },
        "timestamp": "2024-11-15T13:04:12.419178195"
    },
    "sarscov2 - bam - intervals": {
        "content": [
            "d41d8cd98f00b204e9800998ecf8427e",
            [
                "versions.yml:md5,cc6d48cc1e81c73a02d158459281b500"
            ]
        ],
        "meta": {
            "nf-test": "0.9.1",
            "nextflow": "24.10.0"
        },
        "timestamp": "2024-11-15T13:03:36.204251269"
    },
    "sarscov2 - bam - gvcf": {
        "content": [
            "d2ded4648e61277ccc625a5f71624a11",
            [
                "versions.yml:md5,cc6d48cc1e81c73a02d158459281b500"
            ]
        ],
        "meta": {
            "nf-test": "0.9.1",
            "nextflow": "24.10.0"
        },
        "timestamp": "2024-11-15T13:03:59.824887649"
    },
    "sarscov2 - bam - gvcf - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test.g.vcf:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    "versions.yml:md5,cc6d48cc1e81c73a02d158459281b500"
                ],
                "vcf": [
                    [
                        {
                            "id": "test"
                        },
                        "test.g.vcf:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,cc6d48cc1e81c73a02d158459281b500"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.1",
            "nextflow": "24.10.0"
        },
        "timestamp": "2024-11-15T13:04:24.9323125"
    },
    "sarscov2 - bam - no_intervals": {
        "content": [
            "5cf2ef5440f26fc7c2f91a8216cbde04",
            [
                "versions.yml:md5,cc6d48cc1e81c73a02d158459281b500"
            ]
        ],
        "meta": {
            "nf-test": "0.9.1",
            "nextflow": "24.10.0"
        },
        "timestamp": "2024-11-15T13:03:10.486017359"
    }
}
